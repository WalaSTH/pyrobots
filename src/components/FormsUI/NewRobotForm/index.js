// Form input and validation
import * as Yup from "yup";
import { Form, Formik } from "formik";

// Custom components for form input with formik and MUI
import AvatarPreview from "../AvatarPreview";
import FilePreview from "../FilePreview";
import FileUploadInput from "../FileUploadInput";
import FileUploadButton from "../FileUploadButton";
import SubmitFormButton from "../SubmitFormButton";
import TextField from "../TextField";

// MUI components
import { Card, Grid, Typography } from "@mui/material";

// MUI icons
import {
  AddAPhoto as AddAPhotoIcon,
  FileUpload as FileUploadIcon,
  SmartToy as SmartToyIcon,
} from "@mui/icons-material";

const initialFormState = {
  name: "",
  code: "",
  avatar: "",
};

const formValidation = Yup.object().shape({
  name: Yup.string()
    .required("A name is required")
    .min(3, "Must be at least 3 characters long")
    .max(16, "Must be at most 16 characters long")
    .matches(/^[0-9a-z ]+$/i, "Invalid characters"),
  code: Yup.mixed()
    .required("No file selected")
    .test(
      "codeFileFormat",
      "Unsupported file format",
      (value) =>
        value && ["text/x-python", "text/x-python-script"].includes(value.type)
    ),
  avatar: Yup.mixed()
    .notRequired()
    .test(
      "avatarFileFormat",
      "Unsupported file format",
      (value) =>
        !value ||
        (value && ["image/jpeg", "image/png", "image/jpg"].includes(value.type))
    ),
});

export default function NewRobotForm({ onSubmit }) {
  return (
    <Formik
      initialValues={{ ...initialFormState }}
      initialTouched={{
        code: true,
        avatar: true,
      }}
      validationSchema={formValidation}
      onSubmit={onSubmit}
    >
      <Card
        variant="outlined"
        sx={{
          marginTop: 3,
          padding: 3,
          borderRadius: 3,
        }}
      >
        <Form>
          <Grid
            container
            spacing={2}
            sx={{
              display: "flex",
              justifyContent: "center",
            }}
          >
            <Grid
              item
              xs={12}
              sx={{
                textAlign: "center",
              }}
            >
              <Typography component="h1" variant="h5">
                New Robot
              </Typography>
            </Grid>

            <Grid
              item
              xs={12}
              sx={{
                display: "flex",
                justifyContent: "center",
                marginTop: -0.5,
                marginBottom: -0.5,
              }}
            >
              <AvatarPreview
                name="avatar"
                sx={{
                  height: "100px",
                  width: "100px",
                  textAlign: "center",
                }}
              >
                <SmartToyIcon
                  sx={{
                    height: "50px",
                    width: "50px",
                    textAlign: "center",
                  }}
                />
              </AvatarPreview>
            </Grid>

            <Grid
              item
              xs={12}
              sx={{
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
                marginTop: -0.5,
                marginBottom: -0.5,
              }}
            >
              <TextField
                data-testid="robotNameInput"
                name="name"
                label="Name"
                autoComplete="off"
                required
              />
            </Grid>

            <Grid
              item
              xs={12}
              sx={{
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
                marginTop: -0.5,
                marginBottom: -0.5,
              }}
            >
              <FilePreview name="code" />

              <FileUploadInput
                name="code"
                // accept=".py"
                data-testid="robotCodeInput"
              />

              <FileUploadButton
                onClick={() => console.log("corneta")}
                name="code"
                startIcon={<FileUploadIcon />}
              >
                Select file
              </FileUploadButton>
            </Grid>

            <Grid
              item
              xs={12}
              sx={{
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
                marginTop: -0.5,
                marginBottom: -0.5,
              }}
            >
              <FileUploadInput
                name="avatar"
                // accept="image/png,image/jpg,image/jpeg"
                data-testid="robotAvatar"
              />

              <FileUploadButton name="avatar" startIcon={<AddAPhotoIcon />}>
                Select avatar
              </FileUploadButton>
            </Grid>

            <Grid
              item
              xs={12}
              sx={{
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
                marginTop: 0.5,
              }}
            >
              <SubmitFormButton data-testid="submitButton">
                Create robot
              </SubmitFormButton>
            </Grid>
          </Grid>
        </Form>
      </Card>
    </Formik>
  );
}
