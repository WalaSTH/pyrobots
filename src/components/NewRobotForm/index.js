// Form input and validation
import * as Yup from "yup";
import { Form, Formik } from "formik";

// Custom components for form input with formik and MUI
import AvatarPreview from "../FormsUI/AvatarPreview";
import FilePreview from "../FormsUI/FilePreview";
import FileUpload from "../FormsUI/FileUpload";
import SubmitButton from "../FormsUI/SubmitButton";
import TextField from "../FormsUI/TextField";

// MUI components
import { Card, Grid, Typography } from "@mui/material";

// MUI icons
import {
  AddAPhoto as AddAPhotoIcon,
  FileUpload as FileUploadIcon,
  SmartToy as SmartToyIcon,
} from "@mui/icons-material";

const INITIAL_FORM_STATE = {
  name: "",
  code: null,
  avatar: null,
};

const FORM_VALIDATION = Yup.object().shape({
  name: Yup.string()
    .required("A name is required")
    .min(2, "Too short!")
    .max(20, "Too long!")
    .test(
      "isAlphanumeric",
      "Invalid character",
      (value) => value && value.match(/^[0-9a-z]+$/i)
    ),
  code: Yup.mixed()
    .required("No file selected")
    .test(
      "codeFormat",
      "Unsopported file format",
      (value) => value && ["text/x-python"].includes(value.type)
    ),
  avatar: Yup.mixed()
    .notRequired()
    .test(
      "avatarFormat",
      "Unsopported file format",
      (value) =>
        !value ||
        (value && ["image/jpeg", "image/png", "image/jpg"].includes(value.type))
    ),
});

export default function NewRobotForm({ onSubmit }) {
  return (
    <Formik
      initialValues={{ ...INITIAL_FORM_STATE }}
      initialTouched={{
        code: true,
        avatar: true,
      }}
      validationSchema={FORM_VALIDATION}
      onSubmit={onSubmit}
    >
      <Card
        variant="outlined"
        sx={{
          marginTop: 3,
          padding: 3,
          borderRadius: 3,
        }}
      >
        <Form>
          <Grid
            container
            spacing={2}
            sx={{
              display: "flex",
              justifyContent: "center",
            }}
          >
            <Grid
              item
              xs={12}
              sx={{
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
              }}
            >
              <Typography component="h1" variant="h5">
                New Robot
              </Typography>
            </Grid>

            <Grid
              item
              xs={12}
              sx={{
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
              }}
            >
              <AvatarPreview
                name="avatar"
                sx={{
                  height: "100px",
                  width: "100px",
                  textAlign: "center",
                }}
              >
                <SmartToyIcon
                  sx={{
                    height: "50px",
                    width: "50px",
                    textAlign: "center",
                  }}
                />
              </AvatarPreview>
            </Grid>

            <Grid
              item
              xs={12}
              sx={{
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
              }}
            >
              <TextField
                data-testid="robotNameInput"
                name="name"
                label="Robot Name"
                autoComplete="off"
                required
              />

              <FileUpload
                data-testid="robotAvatar"
                accept="image/png,image/jpg,image/jpeg"
                name="avatar"
                buttonProps={{ startIcon: <AddAPhotoIcon /> }}
              >
                Upload avatar
              </FileUpload>
            </Grid>

            <Grid
              item
              xs={12}
              sx={{
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
              }}
            >
              <FilePreview name="code" />

              <FileUpload
                data-testid="robotCodeInput"
                accept=".py"
                name="code"
                buttonProps={{ startIcon: <FileUploadIcon /> }}
              >
                Upload file
              </FileUpload>
            </Grid>

            <Grid
              item
              xs={12}
              sx={{
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
              }}
            >
              <SubmitButton data-testid="submitButton">
                Create robot
              </SubmitButton>
            </Grid>
          </Grid>
        </Form>
      </Card>
    </Formik>
  );
}
