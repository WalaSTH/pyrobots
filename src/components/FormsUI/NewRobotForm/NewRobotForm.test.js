import React from "react";
import "@testing-library/jest-dom/extend-expect";
import renderer, { act } from "react-test-renderer";
import userEvent from "@testing-library/user-event";
import { screen, render } from "@testing-library/react";
import NewRobotForm from "./";

describe("NewRobotForm", () => {
  const user = userEvent.setup();

  it("rendering snapshot", () => {
    const tree = renderer.create(<NewRobotForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Render component", () => {
    act(() => render(<NewRobotForm />));
    screen.getByRole("heading", { name: /New Robot/i });
    screen.getByLabelText(/Name/i);
    screen.getByLabelText(/Select Avatar/i);
    // screen.getByText(/Select Avatar/i);
    screen.getByText(/Select File/i);
    screen.getByText(/Create Robot/i);
  });

  it("checks security of not letting create new robot withou the necessary data", async () => {
    const mockHandler = jest.fn();
    render(<NewRobotForm onSubmit={mockHandler} />);
    const button = screen.getByTestId("submitButton");
    await user.click(button);

    expect(mockHandler).toHaveBeenCalledTimes(0);
  });

  it("correct upload", async () => {
    const mockHandler = jest.fn();
    render(<NewRobotForm onSubmit={mockHandler} />);

    const nameInput = screen.getByTestId("robotNameInput");
    const robotCodeInput = screen.getByLabelText(/Select File/i);
    const file = new File(["hello"], "examplerobot.py", {
      type: "text/x-python",
    });

    await user.type(nameInput, "PatriaoPaper");
    await user.upload(robotCodeInput, file);
  });
});
