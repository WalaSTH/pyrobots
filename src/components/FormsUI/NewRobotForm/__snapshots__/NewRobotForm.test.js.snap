// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewRobotForm rendering snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-xheubk-MuiPaper-root-MuiCard-root"
>
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-9qmdtu-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1fpckdt-MuiGrid-root"
      >
        <h1
          className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          New Robot
        </h1>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-l59uj0-MuiGrid-root"
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular css-1u294kg-MuiAvatar-root"
          color="secondary"
          name="avatar"
          onBlur={[Function]}
          onChange={[Function]}
          value=""
        >
          <img
            className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="none"
          />
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-hzys1r-MuiGrid-root"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          data-testid="robotNameInput"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor=":r0:"
            id=":r0:-label"
          >
            Name
            <span
              aria-hidden={true}
              className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoComplete="off"
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id=":r0:"
              name="name"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              type="text"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-yjsfm1"
              >
                <span>
                  Name
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-hzys1r-MuiGrid-root"
      >
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-1g6u8k6-MuiInputBase-root-MuiInput-root"
          data-testid="robotCodeInput"
          hidden={true}
          onClick={[Function]}
        >
          <input
            accept=".py"
            className="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id="robotCodeInput"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="file"
          />
        </div>
        <label
          className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth css-xsqe0f-MuiButtonBase-root-MuiButton-root"
          htmlFor="robotCodeInput"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="FileUploadIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9 16h6v-6h4l-7-7-7 7h4zm-4 2h14v2H5z"
              />
            </svg>
          </span>
          Select file
        </label>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-hzys1r-MuiGrid-root"
      >
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-1g6u8k6-MuiInputBase-root-MuiInput-root"
          data-testid="robotAvatar"
          hidden={true}
          onClick={[Function]}
        >
          <input
            accept="image/png,image/jpg,image/jpeg"
            className="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id="robotAvatarInput"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="file"
          />
        </div>
        <label
          className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth css-xsqe0f-MuiButtonBase-root-MuiButton-root"
          htmlFor="robotAvatarInput"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AddAPhotoIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 4V1h2v3h3v2H5v3H3V6H0V4h3zm3 6V7h3V4h7l1.83 2H21c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H5c-1.1 0-2-.9-2-2V10h3zm7 9c2.76 0 5-2.24 5-5s-2.24-5-5-5-5 2.24-5 5 2.24 5 5 5zm-3.2-5c0 1.77 1.43 3.2 3.2 3.2s3.2-1.43 3.2-3.2-1.43-3.2-3.2-3.2-3.2 1.43-3.2 3.2z"
              />
            </svg>
          </span>
          Select avatar
        </label>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-srigxs-MuiGrid-root"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth css-189e9pj-MuiButtonBase-root-MuiButton-root"
          data-testid="submitButton"
          disabled={false}
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="submit"
        >
          Create robot
        </button>
      </div>
    </div>
  </form>
</div>
`;
